### PR TITLE
Add `lscpu` Finder

### LINKED ISSUE #57: Add a lscpu -p finder
https://stackoverflow.com/questions/6481005/how-to-obtain-the-number-of-cpus-cores-in-linux-from-the-command-line

### PR DESCRIPTION
**Pull request by thgs**
Closes #57 

Needs some work on the tests still, also not certain for the decorated finder for linux only.

Another interesting option is this:

```
lscpu -J -e=cpu
{
   "cpus": [
      {"cpu":"0"},
      {"cpu":"1"}
   ]
}
```

But cannot tell which one is faster (yet?), pipe to `egrep` or call to `json_decode`. Also wondering still if `egrep` is always installed.

### CODE REVIEW

File: src/Finder/FinderRegistry.php [hunk 1]
```diff
@@ -25,6 +25,9 @@ public static function getAllVariants(): array
             new DummyCpuCoreFinder(1),
             new HwLogicalFinder(),
             new HwPhysicalFinder(),
+            new OnlyOnLinuxFinder(
```
Comment (theofidry) [inline src/Finder/FinderRegistry.php:28]: If we are not sure I would leave is out; but maybe could be tested on a Windows machine with WSL I have no clue. For sure it's not on OSX though

File: src/Finder/LscpuFinder.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace Fidry\CpuCoreCounter\Finder;
+
+use function sprintf;
+
+/**
+ * The number of cores.
```
Comment (theofidry) [inline src/Finder/LscpuFinder.php:19]:
```suggestion
 * The number of logical cores or processors (depending of the configuration).
```

File: src/Finder/LscpuFinder.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    public function __construct(bool $physical = false)
+    {
+        $this->physical = $physical;
+    }
+
+    public function getCommand(): string
+    {
+        return sprintf(
+            "lscpu -p | egrep -v '^#' | %s wc -l",
```
Comment (theofidry) [inline src/Finder/LscpuFinder.php:38]: if this can be done without egrep & wc the better. Any additional parsing (provided the output is not too brittle), should be extracted with `::countCpuCores()`. The reasoning is that the more commands we use the higher the chance that one is not having one of those installed. Whereas parsing strings in PHP, we can do.

There should also be an example or two of how to override the data provider in the test to test more variants

File: src/Finder/LscpuFinder.php [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+            $this->physical ? 'sort -u -t, -k 2,4 |' : ''
+        );
+    }
+
+    public function toString(): string
+    {
+        return 'LscpuFinder';
```
Comment (theofidry) [inline src/Finder/LscpuFinder.php:45]: I would reflect the `physical` property there, you can check `NprocFinder` as a reference

File: src/Finder/OnlyOnLinuxFinder.php [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Fidry\CpuCoreCounter\Finder;
+
+use const PHP_OS_FAMILY;
+
+final class OnlyOnLinuxFinder implements CpuCoreFinder
```
Comment (theofidry) [inline src/Finder/OnlyOnLinuxFinder.php:18]: Might be worth introducing it in a different PR with the other variants `SkipOnLinux`, `OnlyOnMacOs`, `SkipOnMacOs`. Probably an abstract finder could be extracted from it as well. Even if they are not applied in this PR, I think they are good general purpose finders to have

File: src/Finder/FinderRegistry.php [hunk 6]
```diff
@@ -25,6 +25,9 @@ public static function getAllVariants(): array
...
+                new LscpuFinder(),
+            ),
             new _NProcessorFinder(),
             new NProcessorFinder(),
             new NProcFinder(true),
```
Comment (theofidry) [inline src/Finder/FinderRegistry.php:33]: From the `physical` I think it means you can also add it to the `::getDefaultLogicalFinders()` and `::getDefaultPhysicalFinders()` lists

File: src/Finder/FinderRegistry.php (hunk 1 above)
Comment (theofidry) [reply src/Finder/FinderRegistry.php:28]: In any case in the `getAllVariants()` I would not add `OnlyOnLinuxFinder` for sure, but I would add `LscpuFinder` with and without physical.

You can inspect the results in the GA "Inspection"

**Review by theofidry — COMMENTED** [review]
Cool! Thanks for checking it out. There is still one thing or two to figure out but looks pretty good already :)

File: src/Finder/FinderRegistry.php (hunk 1 above)
Comment (thgs) [reply src/Finder/FinderRegistry.php:28]: What about the defaults? Can also separate into a LscpuLogicalFinder and LscpuPhysicalFinder if you think it aligns better with the Hw* finders.

And sorry, what is GA ? I've only run the scripts in bin and the tests.

File: src/Finder/OnlyOnLinuxFinder.php (hunk 5 above)
Comment (thgs) [reply src/Finder/OnlyOnLinuxFinder.php:18]: Yup, sounds good.

File: src/Finder/LscpuFinder.php (hunk 3 above)
Comment (thgs) [reply src/Finder/LscpuFinder.php:38]: After looking around and trying lscpu in what I have available here, it seems to me that simply `-p` will be always supported even on old machines. `-J` was added like 6 years ago and I am not sure about `-p=<desciption>` so Im trying for the "most reliable" `-p`.

**Comment by thgs** [comment]
@theofidry It feels there could be a better implementation for the physical one, not so happy with it yet. The logical one seems simple enough. And I need to add some more test cases, so not ready yet.

File: src/Finder/FinderRegistry.php (hunk 1 above)
Comment (theofidry) [reply src/Finder/FinderRegistry.php]: Sorry I shouldn't be lazy about it, GA=GitHub Actions, e.g. https://github.com/theofidry/cpu-core-counter/actions/runs/3680841567/jobs/6246037753

File: src/Finder/LscpuFinder.php (hunk 3 above)
Comment (theofidry) [reply src/Finder/LscpuFinder.php]: I leave it up to you whether to pick `-J` or not. But I think the rest still applies: you can do the parsing of the output in PHP rather than piping and manipulating it in the command.

File: src/Finder/OnlyOnLinuxFinder.php (hunk 5 above)
Comment (theofidry) [reply src/Finder/OnlyOnLinuxFinder.php]: Revisiting this: maybe a simpler way would be `SkipOnOSFamily` and `OnlyOnOSFamily` where you can pass the `string $osFamily`. It could also contain static methods to make it nicer, `SkipOnOSFamily::onWindows(): self`. The existing Windows variant could then be deprecated in favour of those ones.

File: src/Finder/FinderRegistry.php (hunk 1 above)
Comment (theofidry) [reply src/Finder/FinderRegistry.php]: By the way that also gives you a quick way to check the output of your commands on different platforms, which can help for the unit tests

**Comment by theofidry** [comment]
>It feels there could be a better implementation for the physical one, not so happy with it yet. The logical one seems simple enough. And I need to add some more test cases, so not ready yet.

It would also be fine to introduce the logical one first if you're more confident with that. As long as it's not release it can be freely changed without regards for backward-compatibility too

File: src/Finder/LscpuFinder.php (hunk 3 above)
Comment (thgs) [reply src/Finder/LscpuFinder.php]: Yes, I went for the `compatibility` mode (simple `-p`) which my understanding is that is maintained as BC. Author is not happy about it but for this package I think it's probably the most reliable.

File: src/Finder/LscpuFinder.php (hunk 4 above)
Comment (thgs) [reply src/Finder/LscpuFinder.php]: This has changed with the splitting to two finders.

File: src/Finder/FinderRegistry.php (hunk 6 above)
Comment (thgs) [reply src/Finder/FinderRegistry.php:32]: Added

File: tests/Finder/LscpuPhysicalFinderTest.php [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+
+        yield 'example with unrecognized physical core' => [
+            <<<'EOF'
+# The following is the parsable format, which can be fed to other
+# programs. Each different item in every column has an unique ID
+# starting from zero.
+# CPU,Core,Socket,Node,,L1d,L1i,L2
+0,0,0,0,,0,0,0
+1,-,0,0,,1,1,1
```
Comment (thgs) [inline tests/Finder/LscpuPhysicalFinderTest.php:97]: This is tricky. I am not certain it comes out exactly like this, neither that it means its only one physical core. I am certain lscpu might output `-` on some cases as I can understand from the source code.

Comment (theofidry) [reply tests/Finder/LscpuPhysicalFinderTest.php:97]: checking on docker `ubuntu:rolling`, I get that there:

```
# CPU,Core,Socket,Node
0,0,0,
1,1,0,
2,2,0,
```

So yes I think the output can vary a bit

File: src/Finder/OnlyOnLinuxFinder.php (hunk 5 above)
Comment (theofidry) [reply src/Finder/OnlyOnLinuxFinder.php]: Did that in https://github.com/theofidry/cpu-core-counter/pull/94

Comment (thgs) [reply src/Finder/OnlyOnLinuxFinder.php]: Ok, resolving this.

File: src/Finder/FinderRegistry.php (hunk 1 above)
Comment (thgs) [reply src/Finder/FinderRegistry.php]: Yes it looks pretty good to be honest. There are some errors as I see, which I will address later today (limited time these days).

Comment (thgs) [reply src/Finder/FinderRegistry.php]: I think coding style and e2e should be fixed now.

File: src/Finder/LscpuFinder.php (hunk 2 above)
Comment (thgs) [reply src/Finder/LscpuFinder.php]: Comment above has been split into the two finders. Need any change here or we can resolve?

File: tests/Finder/LscpuPhysicalFinderTest.php (hunk 7 above)
Comment (thgs) [reply tests/Finder/LscpuPhysicalFinderTest.php:96]: Yes, the 1st column will be the logical core and the 2nd column will be the physical so it might be the case of something like:

```
# CPU,Core,Socket,Node
0,0,0,
1,1,0,
2,0,0,
3,1,0,
```

But when for some reason it is not recognized it will output `-`. See line https://github.com/util-linux/util-linux/blame/master/sys-utils/lscpu.c#L303 in the latest `lscpu` that exists for some years. Given that what I say now is correct, currently the code handles this as "unrecognized" and does not count it. But I am not sure how I can get certain 100%. I tried a couple of things and searched the internet for this output but couldn't find it. I suppose I am correct as the man page shows the same

```
-y, --physical
           Display physical IDs for all columns with topology elements (core, socket, etc.). Other than logical IDs, which are assigned by lscpu, physical
           IDs are platform-specific values that are provided by the kernel. Physical IDs are not necessarily unique and they might not be arranged
           sequentially. If the kernel could not retrieve a physical ID for an element lscpu prints the dash (-) character.
```

**Comment by thgs** [comment]
@theofidry hangon, I need to bring this up to date with the latest main branch.

**Comment by thgs** [comment]
@theofidry should be ok now, let me know if you think we should add any additional tests (or anything else).

**Comment by theofidry** [comment]
I have a few nitpicks but I can take care of it. It's more than good enough! Thanks 👏

**Comment by thgs** [comment]
@theofidry Thanks. Really enjoyed the time spent into this. Didn't expect it to be honest.
